File: 572-subtree-of-another-tree/subtree-of-another-tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def isSubtree(self, root, subRoot):
        """
        :type root: TreeNode
        :type subRoot: TreeNode
        :rtype: bool
        """
        def dfs(node):
            if node is None:
                return False
            elif isIdentical(node,subRoot):
                return True
            return dfs(node.right) or dfs(node.left)
        def isIdentical(node1, node2):
            if node1 is None and node2 is None:
                return True
            if node1 is None or node2 is None:
                return False
            return (node1.val == node2.val and 
                    isIdentical(node1.left, node2.left) and 
                    isIdentical(node1.right, node2.right))
        return dfs(root)


        

File: 572-subtree-of-another-tree/subtree-of-another-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean isSubtree(TreeNode root, TreeNode subRoot) {
        if(subRoot==null){
            return true;
        }
        if(root==null){
            return false;
        }

        if(isSameTree(root,subRoot)){
            return true;
        }


        return isSubtree(root.left,subRoot)||isSubtree(root.right,subRoot);
        
    }

    private boolean isSameTree(TreeNode s, TreeNode t){
         if (s == null && t == null) {
            return true;
        }

        // If one is null and the other isn't, they aren't identical
        if (s == null || t == null) {
            return false;
        }

        return (s.val == t.val) && isSameTree(s.left, t.left) && isSameTree(s.right, t.right);
    }
}